### PR TITLE
[NETBEANS-4881] Automatically detect & list Payara Platform version in the server registration panel

### PR DESCRIPTION
**Pull request by jGauravGupta**
This is a feature PR, providing support to parse the Payara Enterprise & Community Server version and dynamically fetch the latest available artifacts of Payara Server from https://repo1.maven.org/maven2/fish/payara/distributions/payara/maven-metadata.xml.

Metadata XML only contains the list of Payara Server Community artifact versions which will be listed (except Beta and Alpha version) in the download combo-box during the registration of the Payara Server in the Apache NetBeans.
image

image

### CODE REVIEW

**Review by lkishalmi — CHANGES_REQUESTED** [review]
Signature file contains a lots of removal. There is no corresponding manifest version change.
I'd suggest to create a new release version for this change.
Maybe a proper consideration splitting the implementation and the interface as an API would be better.
Signatures are only calculated for public packages, so as long as you do not change the interface, you can switch the implementation without the introducing a new release version.

**Comment by jGauravGupta** [comment]
Thanks @lkishalmi for reviewing the PR, I will update the version.

**Comment by jGauravGupta** [comment]
Hi @lkishalmi ,
Can you please restart the job https://travis-ci.org/github/apache/netbeans/jobs/736272562 ?

**Comment by lkishalmi** [comment]
I need a review on this one @matthiasblaesing and @JaroslavTulach for me the sig file changes and the manifest.mf version change does not seem right.

File: enterprise/payara.common/nbproject/org-netbeans-modules-payara-common.sig
Comment (JaroslavTulach) [inline enterprise/payara.common/nbproject/org-netbeans-modules-payara-common.sig:506]: Is this an unnecessary change just adding empty space at the end of the line?

File: enterprise/payara.common/nbproject/org-netbeans-modules-payara-common.sig
Comment (JaroslavTulach) [inline enterprise/payara.common/nbproject/org-netbeans-modules-payara-common.sig:1493]: Changing `enum` to normal `class`? That is binary incompatible change. Now there are two options:
* do the change in a compatible way
* properly document/label/version an incompatible change

In case you want to evolve `enum` to non-enum in a compatible way, read [this essay](http://wiki.apidesign.org/wiki/Enum). It can be done, but just changing to `class ServerDetails` isn't the right approach.

File: enterprise/payara.micro/nbproject/org-netbeans-modules-payara-micro.sig
Comment (JaroslavTulach) [inline enterprise/payara.micro/nbproject/org-netbeans-modules-payara-micro.sig:43]: Removing a field from an API is probably completely unnecessary. Rather mark as `@Deprecated`.

File: enterprise/payara.tooling/nbproject/org-netbeans-modules-payara-tooling.sig
Comment (JaroslavTulach) [inline enterprise/payara.tooling/nbproject/org-netbeans-modules-payara-tooling.sig:1401]: Another attempt to change an `enum` to something extensible. See [again](http://wiki.apidesign.org/wiki/Enum).

File: enterprise/payara.tooling/nbproject/project.xml
Comment (JaroslavTulach) [inline enterprise/payara.tooling/nbproject/project.xml:83]: Removing from list of friends is an incompatible API change.

File: enterprise/payara.tooling/src/org/netbeans/modules/payara/tooling/data/PayaraVersion.java
Comment (JaroslavTulach) [inline enterprise/payara.tooling/src/org/netbeans/modules/payara/tooling/data/PayaraVersion.java:44]: To keep compatibility one would have to:
* keep the `public enum PayaraVersion`
* make it implement `Comparable`
* Introduce new (package private) `class PayaraVersionImproved`
* extract common functionality into `interface PayaraVersionApi`
* make both `enum` and "improved" `class` implement the API interface

File: java/maven.embedder/nbproject/project.xml
Comment (JaroslavTulach) [inline java/maven.embedder/nbproject/project.xml:172]: Expanding list of friends, constitutes an API change and has to be properly versioned. See [friend dependencies](http://wiki.apidesign.org/wiki/FriendDependencies) essay.

**Review by JaroslavTulach — CHANGES_REQUESTED** [review]
The proposed PR contains improperly labeled incompatible changes. If the change is let in in its current form users migrating from previous versions of NetBeans to newer ones using AutoUpdate may experience linkage errors.

One can use NetBeans versioning and module dependencies to prevent update/installation of incompatible module versions. However thinking about it immediately leads to [complex, 3-SAT like solving problems](http://wiki.apidesign.org/wiki/LibraryReExportIsNPComplete) which are usually better to avoid.

Complexities of versioning incompatibilities can be avoided by keeping backward compatibility. In the case of this PR that would mostly mean to [evolve enums in a compatible way](http://wiki.apidesign.org/wiki/Enum), which I'd recommend.

**Comment by lkishalmi** [comment]
@jGauravGupta please address the requested changes by EOB tomorrow (PDT). Otherwise I have to move this PR out of the 12.2 release train.

**Comment by jGauravGupta** [comment]
Thanks for reviewing the PR, I will update the PR soon.

File: enterprise/payara.common/nbproject/org-netbeans-modules-payara-common.sig
Comment (jGauravGupta) [reply enterprise/payara.common/nbproject/org-netbeans-modules-payara-common.sig]: Fixed the formatting issue.

File: enterprise/payara.tooling/nbproject/project.xml
Comment (jGauravGupta) [reply enterprise/payara.tooling/nbproject/project.xml:83]: Payara Modules are originally forked from GlassFish modules, `org.netbeans.modules.payara.cloud` listing is caused by replacing org.netbeans.modules.glassfish -> org.netbeans.modules.payara package, which actually does not exist, so removed it.

File: enterprise/payara.common/nbproject/org-netbeans-modules-payara-common.sig
Comment (JaroslavTulach) [inline enterprise/payara.common/nbproject/org-netbeans-modules-payara-common.sig:1343]: Alas, to be binary backward compatible, you have to keep the old method signature and introduce new method:
```
PayaraPlatformVersionAPI getVersionApi()
```
You can choose the method name, of course. Technically it would even be `getVersion()`, but you are going to have problems to compile Java source that contains two methods without arguments with different return types.

File: enterprise/payara.common/nbproject/org-netbeans-modules-payara-common.sig
Comment (JaroslavTulach) [inline enterprise/payara.common/nbproject/org-netbeans-modules-payara-common.sig:2298]: Keep the original method, introduce new one.

File: enterprise/payara.eecommon/nbproject/org-netbeans-modules-payara-eecommon.sig
Comment (JaroslavTulach) [inline enterprise/payara.eecommon/nbproject/org-netbeans-modules-payara-eecommon.sig:1243]: Keep the original file type. Make the new one (package) `private`!? Encapsulation of fields is always better API choice.

File: enterprise/payara.eecommon/nbproject/org-netbeans-modules-payara-eecommon.sig
Comment (JaroslavTulach) [inline enterprise/payara.eecommon/nbproject/org-netbeans-modules-payara-eecommon.sig:1234]: Keep the original constructor and just introduce another one. The old constructor shall delegate to the new one, I guess.

File: enterprise/payara.eecommon/nbproject/org-netbeans-modules-payara-eecommon.sig
Comment (JaroslavTulach) [inline enterprise/payara.eecommon/nbproject/org-netbeans-modules-payara-eecommon.sig:1268]: Keep old methods, introduce new ones. As the methods differ in parameter type, their overloaded versions shall easily coexist.

File: enterprise/payara.tooling/nbproject/org-netbeans-modules-payara-tooling.sig
Comment (JaroslavTulach) [inline enterprise/payara.tooling/nbproject/org-netbeans-modules-payara-tooling.sig:511]: Keep original and introduce new method. Probably easiest, if you choose different name.

File: enterprise/payara.tooling/nbproject/org-netbeans-modules-payara-tooling.sig
Comment (JaroslavTulach) [inline enterprise/payara.tooling/nbproject/org-netbeans-modules-payara-tooling.sig:1363]: Keep original method, introduce new.

File: enterprise/payara.tooling/nbproject/org-netbeans-modules-payara-tooling.sig
Comment (JaroslavTulach) [inline enterprise/payara.tooling/nbproject/org-netbeans-modules-payara-tooling.sig:1430]: Keep original methods.

File: enterprise/payara.tooling/nbproject/org-netbeans-modules-payara-tooling.sig
Comment (JaroslavTulach) [inline enterprise/payara.tooling/nbproject/org-netbeans-modules-payara-tooling.sig:1616]: Keep and deprecate.

File: enterprise/payara.tooling/nbproject/org-netbeans-modules-payara-tooling.sig
Comment (JaroslavTulach) [inline enterprise/payara.tooling/nbproject/org-netbeans-modules-payara-tooling.sig:2423]: Keep old method, introduce new (probably with different name).

File: enterprise/payara.tooling/nbproject/org-netbeans-modules-payara-tooling.sig
Comment (JaroslavTulach) [inline enterprise/payara.tooling/nbproject/org-netbeans-modules-payara-tooling.sig:1733]: Keep original method.

File: enterprise/payara.tooling/nbproject/org-netbeans-modules-payara-tooling.sig
Comment (JaroslavTulach) [inline enterprise/payara.tooling/nbproject/org-netbeans-modules-payara-tooling.sig:1725]: Keep original methods.

File: enterprise/payara.tooling/nbproject/org-netbeans-modules-payara-tooling.sig
Comment (JaroslavTulach) [inline enterprise/payara.tooling/nbproject/org-netbeans-modules-payara-tooling.sig:1720]: Keep original constructor.

**Review by JaroslavTulach — CHANGES_REQUESTED** [review]
To keep binary compatibility you want to eliminate all removals from the `.sig` files.

**Review by JaroslavTulach — APPROVED** [review]
The current changes in `.sig` files look compatible enough.

**Comment by lkishalmi** [comment]
Due to the sig file change, I'd recommend to merge this after the API snapshot gets merged into master.

**Comment by neilcsmith-net** [comment]
@lkishalmi this is still marked as changes-requested from you, and open against delivery rather than master.  Thoughts?

**Review by lkishalmi — APPROVED** [review]
Approved. Let's have this in. It is up to @neilcsmith-net  whether on master or on delivery.

**Comment by JaroslavTulach** [comment]
So much great work @jGauravGupta and the PR isn't integrated yet!

**Comment by geertjanw** [comment]
Is there anything stopping us from merging this? Unless there are objections, I will merge this on Monday.

**Comment by geertjanw** [comment]
OK, maerging.

**Comment by bdrx312** [comment]
I use netbeans on a private network. After upgrading to netbeans 12.4 I can no longer add a java ee server to my war project due to this change since https://repo1.maven.org/maven2/fish/payara/distributions/payara/maven-metadata.xml is not reachable. Is there a work around?

**Comment by matthiasblaesing** [comment]
@bdrx312 if you really think that this PR is somehow responsible for anything, you need to make a case for that. Most issue reports on that level can be rejected with a "works for me". You need to describe:

- why do you think this PR is responsible
- what do you observe
- how can it be reproduced

If you can't answer these questions, you won't get help.

**Comment by jGauravGupta** [comment]
Hi @bdrx312,

Thanks for your feedback, I will try to reproduce the issue on private network. Update you soon.

**Pull request merged**
